services/requester/pool.go

```diff
@@ -102,6 +154,224 @@ func (t *TxPool) Send(
+//
+// The main advantage of this implementation over the `SingleTxPool`, is the
+// guarantee that transactions originated from the same EOA address, which
+// arrive in a short time interval (about the same as Flow's block productionrate),
+// will be executed in the same order their arrived.
+// This helps to reduce the nonce mismatch errors which mainly occur from the
+// re-ordering of Cadence transactions that happens from Collection nodes.
+type BatchTxPool struct {
+	*SingleTxPool
+	pooledTxs map[gethCommon.Address][]pooledEvmTx
+	txMux     sync.Mutex
+}
+
+var _ TxPool = &BatchTxPool{}
+
+func NewBatchTxPool(
+	client *CrossSporkClient,
+	transactionsPublisher *models.Publisher[*gethTypes.Transaction],
+	logger zerolog.Logger,
+	config config.Config,
+	collector metrics.Collector,
+	keystore *keystore.KeyStore,
+) *BatchTxPool {
+	// initialize the available keys metric since it is only updated when sending a tx
+	collector.AvailableSigningKeys(keystore.AvailableKeys())
+
+	singleTxPool := NewSingleTxPool(
+		client,
+		transactionsPublisher,
+		logger,
+		config,
+		collector,
+		keystore,
+	)
+	batchPool := &BatchTxPool{
+		SingleTxPool: singleTxPool,
+		pooledTxs:    make(map[gethCommon.Address][]pooledEvmTx),
+		txMux:        sync.Mutex{},
+	}
+
+	go batchPool.processPooledTransactions()
+
+	return batchPool
+}
+
+// Add adds the EVM transaction to the tx pool, grouped with the rest of the
+// transactions from the same EOA signer.
+// After the configured `TxBatchInterval`, the collected transations
+// are batched and sent to the Flow network using `EVM.batchRun`, for execution.
+func (t *BatchTxPool) Add(
+	ctx context.Context,
+	tx *gethTypes.Transaction,
+) error {
+	// tx sending should be blocking, so we don't have races when
+	// pooled transactions are being processed in the background.
+	t.txMux.Lock()
+	defer t.txMux.Unlock()
+
+	t.txPublisher.Publish(tx) // publish pending transaction event
+
+	from, err := gethTypes.Sender(gethTypes.LatestSignerForChainID(tx.ChainId()), tx)
+	if err != nil {
+		return fmt.Errorf("failed to derive the sender: %w", err)
+	}
+	txData, err := tx.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	hexEncodedTx, err := cadence.NewString(hex.EncodeToString(txData))
+	if err != nil {
+		return err
+	}
+
+	userTx := pooledEvmTx{txPayload: hexEncodedTx, nonce: tx.Nonce()}
+	t.pooledTxs[from] = append(t.pooledTxs[from], userTx)
+
+	return nil
+}
+
+func (t *BatchTxPool) processPooledTransactions() {
+	for range time.Tick(t.config.TxBatchInterval) {
+		// Take a snapshot here to allow `Add()` continue accepted
+		// incoming EVM transactions, without blocking until the
+		// batch transactions are submitted.
+		snapshot := func() map[gethCommon.Address][]pooledEvmTx {
+			t.txMux.Lock()
+			defer t.txMux.Unlock()
+			copy := t.pooledTxs
+			t.pooledTxs = make(map[gethCommon.Address][]pooledEvmTx)
+			return copy
+		}()
+
+		for address, pooledTxs := range snapshot {
```

Comment:
This syntax is a bit fancy, maybe we can break it:
```
txsGroupedByAccount := snapshot()
for address, pooledTxsFromSameAccount := range txsGroupedByAccount
```

Reply:
Oh sorry, I mean not just renaming the variable name, but also breaking into two statements. 

Because `range snapshot` essentially does two things in a fancy way:
1. call the function to return the map
2. iterate over the map.

`range txsGroupedByAccount` looks like as if `txsGroupedByAccount` is already a map, but actually it's a function.